File: 1.CPU_Memory_Disk_Network/Network/Network_Programming_In_Practice/NPIP_C/t04_local_data_stream/t04_local_data_stream1_server.c
#include <stdio.h>
#include <sys/socket.h>
#include <unistd.h>
#include <string.h>
#include <assert.h>
#include <sys/un.h>

int main() {
  int listen_fd = socket(AF_LOCAL, SOCK_STREAM, 0);
  assert(listen_fd > 0);
  char *local_path = "/tmp/t04_local_data_stream.sock";
  unlink(local_path);

  struct sockaddr_un server_addr, client_addr;
  server_addr.sun_family = AF_LOCAL;
  strcpy(server_addr.sun_path, local_path);

  int bind_res = bind(listen_fd, (struct sockaddr *) &server_addr, sizeof(server_addr));
  assert(bind_res == 0);
  listen(listen_fd, 1024);
  printf("\nServer is listening on %s ...\n", local_path);

  socklen_t client_len = sizeof(client_addr);
  int conn_fd = accept(listen_fd, (struct sockaddr *) &client_addr, &client_len);

  char message[1024];
  while (1) {
    long n = read(conn_fd, message, 1024);
    message[n] = 0;
    printf("Received from client: %s\n", message);

    char send_line[1024];
    sprintf(send_line, "Hi, %s", message);
    write(conn_fd, send_line, strlen(send_line));
  }
}

File: 1.CPU_Memory_Disk_Network/Network/Network_Programming_In_Practice/NPIP_Go/t04_local_datastream/server/T04LocalDataStream1Server.go
package main

import (
	"fmt"
	"net"
	"os"
)

func main() {
	os.Remove("/tmp/t04_local_data_stream_go.sock")
	listenFd, err := net.Listen("unix", "/tmp/t04_local_data_stream_go.sock")
	panicIfError(err, "Listen failed")
	defer listenFd.Close()

	fmt.Println("\nServer is listening on /tmp/t04_local_data_stream_go.sock ...")

	conn, err := listenFd.Accept()
	panicIfError(err, "Accept failed")
	defer conn.Close()

	message := make([]byte, 1024)
	for {
		n, err := conn.Read(message)
		panicIfError(err, "Read failed")
		message[n] = 0
		fmt.Printf("Received from client: %s\n", message)

		sendLine := fmt.Sprintf("Hi, %s", message)
		_, err = conn.Write([]byte(sendLine))
		panicIfError(err)
	}
}

func panicIfError(err error, messages ...any) {
	if err != nil {
		panic(fmt.Sprint(messages, "\nerror: ", err))
	}
}
